### PR TITLE
Xenobiology Gold Slime Spawn Change & Code Cleanup

### PR DESCRIPTION
**Pull request by Hathkar**
Gold slime spawn variables no long double the spawn lists each time the proc is run, and code in general cleaned up a bit. Added a new category to gold slime spawns (NEUTRAL_SPAWN). List is currently empty, but will be filled in a future change. All gold slime reactions now randomly spawn 1-3 mobs. (Previously, water spawned 1 mob, blood spawned 3, and plasma spawned 5)
Credit goes to Cruix for helping with coding.

:cl: Xenobio Gold Slime Spawn Balance & Optimization
add: NEUTRAL_SPAWN list for future changes.
balance: All gold slime reactions now produce 1-3 mobs.
code: Optimized code for less RAM usage.
/:cl:

Rebalances gold slime spawn rates to be more manageable for both antagonist and non-antagonists. Friendly critters (Cats, crabs, corgis, etc.) now spawn more abundantly, and neutral and hostile spawns (mimics, trees, space carp, etc.) spawn at a reduced rate. Assuming xenobiology's slime extractor has been upgraded to T4, there is now a minimum of 4 mobs per slime, and a maximum of 12 per slime. This does not change the in any way the number of slimes produced by xenobiology, only the number of mobs spawned per core. Those breeding gold slimes will still end up with more cores than they can use.

### CODE REVIEW

File: code/__DEFINES/misc.dm
Comment (ExcessiveUseOfCobblestone) [inline code/__DEFINES/misc.dm]: Replace the list into bits so you can add monsters to multiple lists should they apply

File: code/__HELPERS/mobs.dm
Comment (ExcessiveUseOfCobblestone) [inline code/__HELPERS/mobs.dm]: you should check for all three lists

**Review by ExcessiveUseOfCobblestone — CHANGES_REQUESTED** [review]

File: code/__HELPERS/mobs.dm
Comment (ExcessiveUseOfCobblestone) [inline code/__HELPERS/mobs.dm]: If you choose to have bits you'll want to replace this with say

`var/goldcorespawns = initial(SA.gold_core_spawnable)`
`>>>>if(goldcorespawns == NO_SPAWN)`
`>>>>>>>> continue //don't bother `
`>>>>if(goldcorespawns & HOSTILE_SPAWN)`
`>>>>>>>>mob_spawn_meancritters += T`
`>>>>if(goldcorespawns & FRIENDLY_SPAWN)`
`>>>>>>>>mob_spawn_meancritters += T`
`>>>>(same for neutral)`

File: code/__DEFINES/misc.dm
Comment (Hathkar) [reply code/__DEFINES/misc.dm]: Okay, did that change.

File: code/__HELPERS/mobs.dm
Comment (Hathkar) [reply code/__HELPERS/mobs.dm]: Alright, changed this as well.

File: code/__HELPERS/mobs.dm
Comment (Hathkar) [reply code/__HELPERS/mobs.dm]: Okay, amended this as well, I think. Let me know if it's wrong.

**Comment by XDTM** [comment]
By "filled in a future change" you mean in this PR, right? Currently the slime blood reaction, which is likely the most used one, has no mobs.

Also how will you handle mobs that should be both in the neutral and in the hostile pool?

**Comment by Buggy123** [comment]
I don't touch xenobio - how do the new spawn numbers compare to the old ones?

Also A T O M I C. A PR right along the lines of "nerf gold slime cores" got closed _very_ recently, so I'm a little wary of "improves gold slime core code and also nerfs them but a little less".

File: code/__HELPERS/mobs.dm
Comment (DaedalusGame) [reply code/__HELPERS/mobs.dm]: there is, L.len will throw a runtime if L is null.

Comment (Cruix) [reply code/__HELPERS/mobs.dm]: I specifically told him to do one because there will never be a situation where one will be null and the others will not be, because they are all initialized at the same time.

I told him to not do the .len thing because if one of the lists has a length of zero (as the neutral list currently will), it will cause the length of each list to double every time this proc is called, eventually causing the server to run out of memory and crash.

**Comment by Cruix** [comment]
You should absolutely not have three different if statements with duplicate code for each of the lists - as it is now, the first time the proc is called, the lists will all be created three times and discarded twice. There is not reason to not have them all in a single if() looking at just one of the three lists.

**Comment by YakumoChen** [comment]
Again, the whole point of antagging with gold slime cores is that you can spam them three at a time, making a lot of mobs. Like @Buggy123 said, essentially a more severe PR to this just got closed. And to answer his question, the number is "Just 3, every time"

File: code/__HELPERS/mobs.dm
Comment (ExcessiveUseOfCobblestone) [reply code/__HELPERS/mobs.dm]: :up:

**Comment by ExcessiveUseOfCobblestone** [comment]
> Again, the whole point of antagging with gold slime cores is that you can spam them three at a time, making a lot of mobs. Like @Buggy123 said, essentially a more severe PR to this just got closed. And to answer his question, the number is "Just 3, every time"

Uh if you have gold you typically have golems as well.

You can still spam them, it may (probably) will require more cores though as it should.

File: code/__HELPERS/mobs.dm
Comment (Hathkar) [reply code/__HELPERS/mobs.dm]: Alright, amended back to previous version.

**Comment by ExcessiveUseOfCobblestone** [comment]
w-w-w-why did you revert it back to lists instead of bits?

**Comment by Hathkar** [comment]
Oop, sorry. Changed the wrong stuff. I'm still learning how to use Github and code.

**Comment by ExcessiveUseOfCobblestone** [comment]
one sec

**Comment by Hathkar** [comment]
Alright, updated the code a bit more. It compiles cleanly and functions properly in- game.

File: code/__HELPERS/mobs.dm
Comment (Jordie0608) [reply code/__HELPERS/mobs.dm]: Lists are objects, they will always return true when tested; this check does nothing.

Comment (Cruix) [reply code/__HELPERS/mobs.dm]: They will not return true before they are initialized, which is the whole point of initializing them in the if.

**Comment by YakumoChen** [comment]
I still think you should atomize this, we just denied a xenobio nerf a few days ago, putting up another one is disagreeable

**Comment by Buggy123** [comment]
>I still think you should atomize this, we just denied a xenobio nerf a few days ago, putting up another one is disagreeable

**Comment by KorPhaeron** [comment]
Why not nerf stock parts instead of trying to balance around an item that quadruples the effectiveness of something?

**Comment by Buggy123** [comment]
>Why not nerf stock parts instead of trying to balance around an item that quadruples the effectiveness of something?

Look, we've got to leave science _something_.

Unless we get a decent mapper on board with Remove Science 2018.

**Comment by ExcessiveUseOfCobblestone** [comment]
> lowering the mob count from 3 always to 1-3 is the death of science

come on this arms race stuff needs to stop.

**Comment by ExcessiveUseOfCobblestone** [comment]
>Why not nerf stock parts instead of trying to balance around an item that quadruples the effectiveness of something?

@KorPhaeron now that we have them gated by techwebs we need to raise the tech-cost of them significantly considering how much they empower nearly every machine by folds.

**Comment by Hathkar** [comment]
> I still think you should atomize this, we just denied a xenobio nerf a few days ago, putting up another one is disagreeable

And what does atmomizing mean exactly? I'm fine with a WIP and/or Do Not Merge Tag being applied for a while until I flesh out the Neutral spawn list.

**Comment by KorPhaeron** [comment]
>X item and its acquisition is fine for years

>golfball and raz add insane machine parts that triple or quadruple the availability or effectiveness of that item

>nerf X item or make it harder to obtain normally because people are getting too many too quickly so now you can get a ton of them but they aren't worth getting

Another classic /tg/ balance saga

**Comment by Buggy123** [comment]
>And what does atmomizing mean exactly? I'm fine with a WIP and/or Do Not Merge Tag being applied for a while until I flesh out the Neutral spawn list.

Put the balance changes in one PR, put the code fixes in another PR. If you combine them then a maintainer, if they don't like the balance change, is forced to weigh between merging it because of the code fixes, or denying it because of the unwanted balance change.

File: code/__HELPERS/mobs.dm
Comment (ExcessiveUseOfCobblestone) [reply code/__HELPERS/mobs.dm]: Will the list actually be static when you put it in a proc or will it generate a new list everytime this proc is called.

Comment (Cruix) [reply code/__HELPERS/mobs.dm]: The list is actually static but scoped to only this proc.

**Comment by ExcessiveUseOfCobblestone** [comment]
@KorPhaeron it's just a var edit to keep them in the game as extreme upgrades with techwebs now tbf 😮

**Comment by Buggy123** [comment]
I told you man, I TOLD you about atomizing!

**Comment by Hathkar** [comment]
Could I at least ask why this was closed?

**Pull request closed without merging**
